system: Provide possibility to install PCP

larskarlitski opened the pull request.

martinpitt commented on pkg/systemd/host.js:
```diff
@@ -361,9 +362,11 @@ PageServer.prototype = {
         function refresh_pmlogger_state() {
-            if (!pmlogger_service.exists)
+            if (!pmlogger_service.exists) {
+                $('#system-information-enable-pcp').show();
```
This should first check if a `pcp` package is actually available. Our packagekit library has `check_missing_packages()` for that, which should be suitable. We don't want to show this on debian-stable, as it would just fail.

mvollmer replied:
I was thinking that this is determined statically, per distro, when building.  E.g., on Fedoras we install nfs-utils for NFS client support, on Debian we install nfs-common.  On other distros that haven't made a choice in their packaging, we don't offer to install anything.

`check_missing_packages` is used internally by the install dialog to figure out what the dependencies are and how much needs to be downloaded.  There is no point in running it before opening the dialog, I would say.

martinpitt replied:
OK, so a new `configure` option, like for NFS? Makes sense.

martinpitt commented on test/verify/check-system-info:
```diff
@@ -380,5 +380,27 @@ class TestSystemInfo(MachineCase):
+        # Don't install/uninstall pcp in this test, so that we avoid hitting
+        # the network. Instead, check that the link to install pcp is there
+        # when it is not installed.
+        installed = m.execute("systemctl show -p LoadState pmlogger") == "LoadState=loaded"
```
TBH I'd prefer having a static list of OSes here. I. e. at least "debian-stable", I'm not sure about the Atomics. Every other OS ought to already have it installed.

So this should check that `Enable...` is *not* shown on any OS by default, as all our images which *can* have pcp now do (i. e. all but debian-stable). Then everywhere but debian-stable you can remove the pcp package, reload the page, and check that you now get the `Enable...` offered. This doesn't need networking.

larskarlitski commented:
Thanks for the review. Rebased, added a new configure option akin to `--with-nfs-client-packages`, and changed the test according to @martinpitt's suggestion.

larskarlitski commented:
Update: use `which pmlogger` to more robustly figure out if pcp is installed; use `pkcon` to remove pcp on centos-7 in the tests, because packagekit holds the yum lock while it is running; reload the browser for some images.

martinpitt commented on configure.ac:
```diff
@@ -381,6 +381,13 @@ AC_ARG_WITH(nfs-client-package,
 
+AC_ARG_WITH(pcp-package,
+            AC_HELP_STRING([--with-pcp-package=JSON],
+                           [Package that is necessary for PCP, or 'false']),
```
Why does this have to be a config option? It's the same name in all OSes that we support, and unlike what the NFS tests do, your tests don't even need a fake package name.

larskarlitski replied:
You and @mvollmer came to this conclusion in the last round of review, no? Maybe I misunderstood?

martinpitt replied:
I don't veto it, and we can do it for the sake of keeping the same pattern. I was just thinking that it's overkill and even harmfull to make our own package name ("cockpit-pcp") configurable, as we control that ourselves. That of course depends on the answer to my question below, whether we should really only install "pcp" or "cockpit-pcp".

martinpitt commented on tools/cockpit.spec:
```diff
@@ -151,6 +151,7 @@ exec 2>&1
     %{?rhel:--without-storaged-iscsi-sessions} \
     --with-appstream-data-packages='[ "appstream-data" ]' \
     --with-nfs-client-package='"nfs-utils"' \
+    --with-pcp-package='"pcp"' \
```
Shouldn't we install cockpit-pcp here, not just pcp?

martinpitt commented on test/verify/check-system-info:
```diff
@@ -380,5 +380,43 @@ class TestSystemInfo(MachineCase):
         b.enter_page("/system")
 
 
+    @skipImage("PCP config not yet in manifest and no packagekit", "rhel-7-5")
```
This is misleading - rhel-7-5 does have PackageKit. It's just that the old cockpit-system on rhel-7-5 does not have the install-on-demand feature.

larskarlitski replied:
Changed, thanks.

martinpitt commented on test/verify/check-system-info:
```diff
@@ -380,5 +380,43 @@ class TestSystemInfo(MachineCase):
+            elif m.image.startswith("rhel-7"):
+                m.execute("yum remove -y pcp")
+            else:
+                raise NotImplementedError("unknown image " + m.image)
```
Couldn't this always use `pkcon remove` on all OSes?

larskarlitski replied:
Of course. Perfect.

martinpitt commented on pkg/systemd/host.js:
```diff
@@ -334,43 +335,66 @@ PageServer.prototype = {
+            /* HACK: The pcp packages on Ubuntu and Debian inclue SysV init
+             * scripts in /etc, which stay around when removing (as opposed to
+             * purging) the package. This makes systemd think that there's
+             * still a unit there, and pmlogger_service.exists will be true.
```
Clarification: This is not a "thinko" of systemd -- there *is* a valid and enabled SysV init script, and thus also a unit. There is no way/API to tell about a generic init.d script whether it is "backed" by a package or not.

larskarlitski replied:
Makes sense. Rephrased.

martinpitt commented on pkg/systemd/host.js:
```diff
@@ -334,10 +335,40 @@ PageServer.prototype = {
+            /* HACK: The pcp packages on Ubuntu and Debian inclue SysV init
+             * scripts in /etc, which stay around when removing (as opposed to
+             * purging) the package. Systemd treats those as valid units, even
+             * if they're not backed by packages anymore.  Thus, and
```
Just in case you have to push again: please unbreak this "Thus, and" sentence.

larskarlitski replied:
Oops, thanks.

martinpitt commented on pkg/systemd/host.js:
```diff
@@ -334,10 +335,40 @@ PageServer.prototype = {
+                        $('#server-pmlogger-onoff-row').hide();
+                    } else if (!pmlogger_promise) {
+                        $('#system-information-enable-pcp').hide();
+                        console.log(pmlogger_service);
```
Does this even show something sensible? More than just `[Object]`?

larskarlitski replied:
It doesn't. I used it while debugging an forgot to take it out.

martinpitt commented on test/verify/check-system-info:
```diff
@@ -380,5 +380,34 @@ class TestSystemInfo(MachineCase):
+            # remove the cockpit-pcp package
+            m.execute("pkcon remove -y pcp")
+
+        if m.image in ["centos-7", "rhel-7"] or m.image.startswith("debian") or m.image.startswith("ubuntu"):
```
This needs a comment. IIRC this is because these OSes ship an init.d script which isn't getting removed along with the package?

martinpitt commented on test/verify/check-system-info:
```diff
@@ -380,5 +380,34 @@ class TestSystemInfo(MachineCase):
+        b.wait_in_text("#system-information-enable-pcp", "Enable persistent metrics…")
+        b.click("#system-information-enable-pcp-link")
+        b.wait_present("#cockpit_modal_dialog")
+        b.wait_in_text("#cockpit_modal_dialog .modal-header", "Install Software")
```
Oh right, this would fail because the VMs are offline. You could use test/verify/packagelib.py, derive your test from `PackageCase`, and `createPackage()` a fake `cockpit-pcp` to test it all the way through. See `class TestStoragePackages` as an example, I'm happy to walk you through it.

martinpitt commented on test/verify/check-system-info:
```diff
@@ -380,5 +381,48 @@ class TestSystemInfo(MachineCase):
+
+        pmlogger = {
+            "/usr/bin/pmlogger": "",
+            "/etc/systemd/system/pmlogger.service": "[Service]\nExecStart=/usr/sbin/true"
```
This should be /bin/true, this path does not exist.

larskarlitski replied:
Ugh, how did that happen :roll_eyes: Sorry

martinpitt commented on test/verify/check-system-info:
```diff
@@ -380,5 +381,48 @@ class TestSystemInfo(MachineCase):
+        m.execute("pkcon install -y cockpit-pcp")
+
+        b.wait_visible("#server-pmlogger-onoff-row")
+        b.wait_not_visible("#system-information-enable-pcp")
```
I still don't understand this test case. The above opens the dialog, but never acks the package install, and this now installs the fake package manually. Shouldn't the fake package be created at the top of the test case, and then you let the actual cockpit code install the package all the way through?

larskarlitski replied:
You're right. I've gone through a lot of changes on this one, and this is a holdout from before I had the mock package. I've changed it to install by clicking the button instead of doing it through `pkcon`. Thanks!

larskarlitski commented:
> The test case fails, and it exposes a layout bug: The label gets cut off. See a screenshot from any failed test.

Ah, as race: the test didn't wait for the "Install" button to become enabled. Fixed the layout bug by overriding `whitespace: normal` for the link (same as for the operating system and hardware text).

larskarlitski commented:
Pushed another fix: use `wait_not_present` instead of `wait_not_visible` for the dialog.

martinpitt commented:
Screenshot for release notes:

![pcp-install-on-demand](https://user-images.githubusercontent.com/200109/42575289-62fae9ba-8520-11e8-8e16-1794606458b4.png)
